rfcs/theme-shape.md

```diff
@@ -0,0 +1,228 @@
+# RFC: Update theme shape
+
+@miroslavstastny, @layershifter
+
+_This RFC updates [theme-tokens.md](./theme-tokens.md) RFC. Once this on is approved, we should merge the two RFCs together._
+
+## Summary
+
+As described in [theme-tokens.md](./theme-tokens.md) the current theme consists of global and alias tokens.
+All tokens are injected into DOM as CSS variables, resulting in 1200 CSS variables. With this approach, when a theme switches, there is no need to re-compute any component styles, by just setting the new token values to the CSS variables, all components change their appearance accordingly.
+
+However, 1200 CSS variables in DOM affects browser performance. This RFC proposes ways how to improve theme performance.
+
+## <a name="performance-analysis"></a>Performance analysis
+
+The current theme consists of ~1200 tokens which are injected into DOM as CSS variables. It impacts not only Javascript performance but also browser performance in styles computation phase.
+
+To see the performance impact we render 20 `FluentProvider` components side by side (**each** injecting a class with 1200 CSS variables) - mount and unmount 10 times and measure the performance in Chrome profiler with 6x CPU slowdown. After each mount, a CSS attribute is accessed to force reflow in order to be able to measure the rendering performance.
```

Comment:
How aligned with the common case of using Fluent this is: 
- Won't there usually be a single ThemeProvider at the root for an application and maybe 1 or 2 sub ThemeProviders in more advanced scenarios?  
- Won't the css variables be set on the theme root and cascade down to the components?
- If the first two answers are yes, then would the common case be that we are only updating a single point in the DOM with the 1200 variable value changes?
-

Reply:
Totally correct - just one (2 in an edge case) `ThemeProvider`.

But even just that one provider makes **20ms** difference between the **Master** and **Inline values to alias tokens** in a real partner application and these 20ms is a no-go for them.

This test is an attempt to have an isolated perf example which we can run on the library side (including the CI) before testing it on partner side.
_(That being said, only the **Master** vs **Inline values to alias tokens** experiment has been tested in the partner application so far as we do not have a real implementation for the other steps)._